### PR TITLE
Pass Severity instead of String to Logger::Formatter

### LINKED ISSUE #4355: Shouldn't Logger#formatter take a Severity instead of a String?
I was surprised `Logger#formatter` takes a `String` for its first parameter `severity` instead of a `Logger::Severity` enum. Is this on purpose? Why should I write:

```crystal
LOG = Logger.new(STDOUT)
LOG.formatter = Logger::Formatter.new do |severity, datetime, progname, message, io|
  io << case severity
  when "DEBUG"
    ":: "
  when "INFO"
    ">> "
  end
  io << message
end
```

instead of:

```crystal
LOG = Logger.new(STDOUT)
LOG.formatter = Logger::Formatter.new do |severity, datetime, progname, message, io|
  io << case severity
  when .debug?
    ":: "
  when .info?
    ">> "
  end
  io << message
end
```

### PR DESCRIPTION
**Pull request by Sija**
Fixes #4355

### CODE REVIEW

File: src/logger.cr [hunk 1]
```diff
@@ -58,11 +58,12 @@ class Logger
     UNKNOWN
   end
 
-  alias Formatter = String, Time, String, String, IO ->
+  alias Formatter = Severity, Time, String, String, IO ->
 
   private DEFAULT_FORMATTER = Formatter.new do |severity, datetime, progname, message, io|
-    io << severity[0] << ", [" << datetime << " #" << Process.pid << "] "
-    io << severity.rjust(5) << " -- " << progname << ": " << message
+    label = severity.unknown? ? "ANY" : severity.to_s
```
Comment (mverzilli) [inline src/logger.cr]: Why not just use "U" (for "Unknown")?

Comment (Sija) [inline src/logger.cr]: I just copy-pasted this line from previous version. Why not just `severity.to_s` without branching into other symbols?

Comment (mverzilli) [inline src/logger.cr]: Hadn't noticed it was already like that. Then let's just leave it like this.

**Comment by Sija** [comment]
@mverzilli done.

**Comment by mverzilli** [comment]
@Sija I meant to document the `formatter` property, which is public. What do you think about this?

1. Remove that comment from the `alias` definition
2. Add this on top of the `formatter` property:
```
# Customizable `Proc` (with a reasonable default)
# which the Logger uses to format and print its entries.
#
# Use this setter to provide a custom formatter.
# The Logger will invoke it with the following arguments:
#  - severity: a `Logger::Severity`.
#  - datetime: `Time`, the entry's timestamp
#  - progname: `String`, the program name, if set when the logger was built
#  - message: `String`, the body of a message
#  - io: `IO`, the Logger's stream, to which you must write the output
# you want.
#
# Example:
#
# ```
# require "logger"
#
# logger = Logger.new(STDOUT)
# logger.progname = "YodaBot"
#
# logger.formatter = Logger::Formatter.new do |severity, datetime, progname, message, io|
#   label = severity.unknown? ? "ANY" : severity.to_s
#   io << label[0] << ", [" << datetime << " #" << Process.pid << "] "
#   io << label.rjust(5) << " -- " << progname << ": " << message
# end
#
# logger.warn("Fear leads to anger. Anger leads to hate. Hate leads to suffering.")
#
# # Prints to the console:
# # "W, [2017-05-06 18:00:41 -0300 #11927]  WARN --
# #  YodaBot: Fear leads to anger. Anger leads to hate. Hate leads to suffering."
# ```
```

Btw, I would have liked to directly add a commit to the PR without bothering you, but I'm not sure what the best workflow for doing it is :(.

**Comment by RX14** [comment]
@mverzilli I think maintainers can push directly to most PRs now.

**Comment by Sija** [comment]
I've force-pushed the last commit with your docs added. In the future it's fine with me with you'd do it yourself. I've took a liberty to modify some interpunction and wording here & there if that's ok with you.

**Comment by Sija** [comment]
image
@RX14 with above option checked, they do.

**Comment by RX14** [comment]
@Sija that box is checked by default now though, so I imagine most prs don't bother to change it.

**Comment by mverzilli** [comment]
Nice, it does work :).

**Pull request merged**